### PR TITLE
Process withdrawal

### PR DESCRIPTION
**Pull request by Redidacove**
What type of PR is this?
This is a helper function implementation pr for processing withdrawals

What does this PR do? Why is it needed?
This Pr implements the process_withdrawals fn acc to the eip7732 specs implementation

### CODE REVIEW

File: beacon-chain/core/blocks/withdrawals.go [hunk 1]
```diff
@@ -148,40 +148,58 @@ func ValidateBLSToExecutionChange(st state.ReadOnlyBeaconState, signed *ethpb.Si
 //	        next_validator_index = ValidatorIndex(next_index % len(state.validators))
 //	        state.next_withdrawal_validator_index = next_validator_index
 func ProcessWithdrawals(st state.BeaconState, executionData interfaces.ExecutionData) (state.BeaconState, error) {
-	expectedWithdrawals, partialWithdrawalsCount, err := st.ExpectedWithdrawals()
```
Comment (potuz) [inline beacon-chain/core/blocks/withdrawals.go:151]: The godoc needs to be updated to ePBS.

File: beacon-chain/core/blocks/withdrawals.go [hunk 2]
```diff
@@ -148,40 +148,58 @@ func ValidateBLSToExecutionChange(st state.ReadOnlyBeaconState, signed *ethpb.Si
...
+			return nil, errors.Wrap(err, "could not set withdrawals root")
+		}
+	} else {
+		var wdRoot [32]byte
+		if executionData.IsBlinded() {
+			r, err := executionData.WithdrawalsRoot()
+			if err != nil {
+				return nil, errors.Wrap(err, "could not get withdrawals root")
+			}
+			wdRoot = bytesutil.ToBytes32(r)
```
Comment (potuz) [inline beacon-chain/core/blocks/withdrawals.go:184]:
```suggestion
			copy(wdRoot[:], r)
```

File: beacon-chain/core/blocks/withdrawals_test.go [hunk 3]
```diff
@@ -1028,6 +1030,23 @@ func TestProcessWithdrawals(t *testing.T) {
...
+				FullWithdrawalIndices:        []primitives.ValidatorIndex{7, 19, 28, 1},
+				Withdrawals: []*enginev1.Withdrawal{
+					fullWithdrawal(7, 22), fullWithdrawal(19, 23), fullWithdrawal(28, 25),
+				},
+				LatestExecutionPayloadHeader: &enginev1.ExecutionPayloadHeaderEPBS{
+					BlockHash: bytesutil.Bytes32(2),
+				},
+			},
+			Control: control{
+				ExpectedError: true,
```
Comment (potuz) [inline beacon-chain/core/blocks/withdrawals_test.go:1047]: Why an error here if the parent node is not full we don't return an error.

File: beacon-chain/core/blocks/withdrawals.go [hunk 4]
```diff
@@ -121,9 +120,16 @@ func ValidateBLSToExecutionChange(st state.ReadOnlyBeaconState, signed *ethpb.Si
 // Spec pseudocode definition:
 //
 //	def process_withdrawals(state: BeaconState, payload: ExecutionPayload) -> None:
+//		if st.version() >= version.EPBS :
+//			if not is_parent_block_full(state): # [New in EPBS:EIP7732]
+//		 		return
+//
```
Comment (potuz) [inline beacon-chain/core/blocks/withdrawals.go:126]: This does not match the python declaration

File: beacon-chain/core/blocks/withdrawals.go [hunk 5]
```diff
@@ -216,6 +240,7 @@ func ProcessWithdrawals(st state.BeaconState, executionData interfaces.Execution
 			nextValidatorIndex = 0
 		}
 	}
+
```
Comment (potuz) [inline beacon-chain/core/blocks/withdrawals.go:243]:
```suggestion
```

File: beacon-chain/core/blocks/withdrawals_test.go [hunk 6]
```diff
@@ -12,11 +12,13 @@ import (
 	fieldparams "github.com/prysmaticlabs/prysm/v5/config/fieldparams"
 	"github.com/prysmaticlabs/prysm/v5/config/params"
 	consensusblocks "github.com/prysmaticlabs/prysm/v5/consensus-types/blocks"
+	"github.com/prysmaticlabs/prysm/v5/consensus-types/epbs"
 	"github.com/prysmaticlabs/prysm/v5/consensus-types/interfaces"
 	"github.com/prysmaticlabs/prysm/v5/consensus-types/primitives"
 	"github.com/prysmaticlabs/prysm/v5/crypto/bls"
 	"github.com/prysmaticlabs/prysm/v5/crypto/bls/common"
 	"github.com/prysmaticlabs/prysm/v5/crypto/hash"
+
```
Comment (potuz) [inline beacon-chain/core/blocks/withdrawals_test.go:21]:
```suggestion
```

**Review by terencechain — CHANGES_REQUESTED** [review]
I dont think [df7b205](https://github.com/prysmaticlabs/prysm/pull/14297/commits/df7b2057279e5cdd77c0a52276a901b8734df42d) is correct, instead of duplicating the entire test for test setup, you could have just initialized the field directly. Adding this fixes the panic for me:

Screenshot 2024-08-13 at 7 11 32 AM

**Pull request merged**